File: ios/ExpoWalletCoreModule.swift
import ExpoModulesCore
import WalletCore
import UIKit
import SwiftProtobuf

public class ExpoWalletCoreModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoWalletCore")

    Function("create") { () -> [String : String] in
      let wallet = HDWallet(strength: 128, passphrase: "")!
      return [
        "entropy": wallet.entropy.hexString,
        "mnemonic": wallet.mnemonic,
        "seed": wallet.seed.hexString
      ]
    }

    Function("createWithMnemonic") { (mnemonic: String) -> [String : String] in
      let wallet = HDWallet(mnemonic: mnemonic, passphrase: "")!
      return [
        "entropy": wallet.entropy.hexString,
        "mnemonic": wallet.mnemonic,
        "seed": wallet.seed.hexString
      ]
    }

    Function("getAddressForCoin") { (mnemonic: String, coin: UInt32) -> String in
      let wallet = HDWallet(mnemonic: mnemonic, passphrase: "")!
      let coin = CoinType(rawValue: coin)!
      return wallet.getAddressForCoin(coin: coin)
    }

    Function("getKeyForCoin") { (mnemonic: String, coin: UInt32) -> String in
      let wallet = HDWallet(mnemonic: mnemonic, passphrase: "")!
      let coin = CoinType(rawValue: coin)!
      return wallet.getKeyForCoin(coin: coin).data.hexString
    }

    Function("sign") { (data: String, coin: UInt32) -> String in
        let coin = CoinType(rawValue: coin)!
        let signerOutput = AnySigner.nativeSign(data: Data(hexString: data)!, coin: coin)
        return signerOutput.hexString
    }
  }
}

File: android/src/main/java/expo/modules/walletcore/ExpoWalletCoreModule.kt
package expo.modules.walletcore

import com.google.protobuf.ByteString
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Ethereum
import java.math.BigInteger

fun ByteArray.toHexString() : String {
  return this.joinToString("") {
    java.lang.String.format("%02x", it)
  }
}

class ExpoWalletCoreModule : Module() {
  init {
    System.loadLibrary("TrustWalletCore")
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoWalletCore")

    Function("create") {
      val wallet = HDWallet(128, "")
      mapOf(
        "entropy" to wallet.entropy().toString(),
        "mnemonic" to wallet.mnemonic(),
        "seed" to wallet.seed().toString()
      )
    }

    Function("createWithMnemonic") { mnemonic: String ->
      val wallet = HDWallet(mnemonic, "")
      mapOf(
        "entropy" to wallet.entropy().toString(),
        "mnemonic" to wallet.mnemonic(),
        "seed" to wallet.seed().toString()
      )
    }

    Function("getAddressForCoin") { mnemonic: String, coin: Int ->
      val wallet = HDWallet(mnemonic, "")
      wallet.getAddressForCoin(CoinType.createFromValue(coin))
    }

    Function("getKeyForCoin") { mnemonic: String, coin: Int ->
      val wallet = HDWallet(mnemonic, "")
      wallet.getKeyForCoin(CoinType.createFromValue(coin)).data().toHexString()
    }

    Function("sign") { data: String, coin: Int ->
      val output = AnySigner.nativeSign(BigInteger(data, 16).toByteArray(), coin)
      output.toHexString()
    }
  }
}
